### PR TITLE
feat: integrate sglang in camel

### LINKED ISSUE #923: [Feature Request] Integrate `SGlang`
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [ ] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

Integrate `SGlang` for fast serving.

https://github.com/sgl-project/sglang

### Solution

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by NeilJohnson0930**
## Description

Describe your changes in detail.

## Motivation and Context

 `close #923 `

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/configs/sglang_config.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    json_schema: Optional[str] = None
+    min_new_tokens: int = 0
+    top_k: int = -1
+    min_p: float = 0.0
+    repetition_penalty: float = 1.0
+    logprob_start_len: Optional[Union[Sequence[int], int]] = None
+    top_logprobs_num: Optional[Union[Sequence[int], int]] = None
+    return_text_in_logprobs: bool = False
+    stream: bool = False
+    max_tokens: Union[int, NotGiven] = NOT_GIVEN
```
Comment (raywhoelse) [inline camel/configs/sglang_config.py:117]: From my personal point of view, there is no need to set so many parameters. You only need to set some commonly used parameters. For details, please refer to ollama

Comment (raywhoelse) [reply camel/configs/sglang_config.py:117]: @Wendong-Fan

Comment (Wendong-Fan) [reply camel/configs/sglang_config.py:117]: Thanks @raywhoelse , I believe it's fine to include as many parameters as possible; users are free to choose and use any of them as needed

File: camel/configs/__init__.py [hunk 2]
```diff
@@ -29,6 +29,7 @@
     SambaCloudAPIConfig,
     SambaVerseAPIConfig,
 )
+from .sglang_config import SGLANG_API_PARAMS, SGLANGConfig
```
Comment (Wendong-Fan) [inline camel/configs/__init__.py:32]:
```suggestion
from .sglang_config import SGLANG_API_PARAMS, SGLangConfig
```

File: camel/models/sglang_model.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            self._url = "http://localhost:30000/v1"
+            print(
+                f"Ollama server started on {self._url} "
+                f"for {self.model_type} model."
+            )
+        except Exception as e:
+            print(f"Failed to start sglang server: {e}.")
```
Comment (Wendong-Fan) [inline camel/models/sglang_model.py:93]: use SGLang native feature as we discussed

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @NeilJohnson0930 @raywhoelse ! Overall LGTM, left some comments below

File: camel/models/sglang_model.py (hunk 3 above)
Comment (zhaochenyang20) [reply camel/models/sglang_model.py:93]: @Wendong-Fan @liuxukun2000 

Hello. This is Chenyang from SGLang team. Regarding the launch SGLang engine, you can try these two ways:

```python
popen_launch_server(
            cls.model, cls.base_url, timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH
        )
```

This is from https://github.com/sgl-project/sglang/blob/main/test/srt/test_update_weights_from_disk.py

And, alternatively, you can try this:

```python
from sglang.utils import (
    execute_shell_command,
    wait_for_server,
    terminate_process,
    print_highlight,
)

server_process = execute_shell_command(
    "python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --port 30000 --host 0.0.0.0"
)

wait_for_server("http://localhost:30000")
```

This is from our docs:

https://sgl-project.github.io/backend/openai_api_completions.html

File: camel/configs/sglang_config.py (hunk 1 above)
Comment (zhaochenyang20) [reply camel/configs/sglang_config.py:117]: Also, SGLang's default parameter is generally good enough. Do not need to set it too complicated.

**Comment by zhaochenyang20** [comment]
Also, we have some additional comments:

1. Could you try to submit an issue to SGLang regarding the desired input and output of func calling in SGLang? Func calling is highly specific to each model, we may want to see the expected behavior of Llama/Qwen.

2. Could you please try to benchmark the speed and accuracy of SGLang compared with vllm? SGLang should be as good as vllm but steadily faster. If you find this is not as expected, please do let me know.

@NeilJohnson0930 @Wendong-Fan @lightaime

**Comment by zhaochenyang20** [comment]
@NeilJohnson0930 Hello. Could also provide these contents?

> Could you please try to benchmark the speed and accuracy of SGLang compared with vllm? SGLang should be as good as vllm but steadily faster. If you find this is not as expected, please do let me know.

**Comment by Wendong-Fan** [comment]
> Also, we have some additional comments:
> 
> 1. Could you try to submit an issue to SGLang regarding the desired input and output of func calling in SGLang? Func calling is highly specific to each model, we may want to see the expected behavior of Llama/Qwen.
> 2. Could you please try to benchmark the speed and accuracy of SGLang compared with vllm? SGLang should be as good as vllm but steadily faster. If you find this is not as expected, please do let me know.
> 
> @NeilJohnson0930 @Wendong-Fan @lightaime

Thanks @zhaochenyang20!
The open-source model tool-calling feature is supported in [this implementation](https://github.com/camel-ai/camel/blob/chat_agent_refactor_wd/camel/agents/chat_agent.py#L722). Essentially, we use a prompt-based approach to enable tool-calling across all models. This functionality is managed on the CAMEL side, so there are no additional requirements for SGLang at this time.

Regarding the comparison of model speed and accuracy, could @NeilJohnson0930 assist with conducting a simple test using different frameworks for the same open-source model?

**Review by Wendong-Fan — APPROVED** [review]

**Comment by zhaochenyang20** [comment]
> > Also, we have some additional comments:
> > 
> > 1. Could you try to submit an issue to SGLang regarding the desired input and output of func calling in SGLang? Func calling is highly specific to each model, we may want to see the expected behavior of Llama/Qwen.
> > 2. Could you please try to benchmark the speed and accuracy of SGLang compared with vllm? SGLang should be as good as vllm but steadily faster. If you find this is not as expected, please do let me know.
> > 
> > @NeilJohnson0930 @Wendong-Fan @lightaime
> 
> Thanks @zhaochenyang20! The open-source model tool-calling feature is supported in [this implementation](https://github.com/camel-ai/camel/blob/chat_agent_refactor_wd/camel/agents/chat_agent.py#L722). Essentially, we use a prompt-based approach to enable tool-calling across all models. This functionality is managed on the CAMEL side, so there are no additional requirements for SGLang at this time.
> 
> Regarding the comparison of model speed and accuracy, could @NeilJohnson0930 assist with conducting a simple test using different frameworks for the same open-source model?

Thanks! Love to see your resuts.

**Review by raywhoelse — APPROVED** [review]

**Comment by NeilJohnson0930** [comment]
Hi @zhaochenyang20, You can see the speed of SGLang compared with Vllm in the url link below:
https://github.com/camel-ai/camel/pull/1317/files

**Comment by lightaime** [comment]
I think the token per second is not correct for some reason though

**Comment by Wendong-Fan** [comment]
> I think the token per second is not correct for some reason though

@NeilJohnson0930  will redo the experiment

**Comment by zhaochenyang20** [comment]
@NeilJohnson0930 @Wendong-Fan I feel it too strange for vllm. What's the serving parameters you used? Also, how about the accuracy results? Thanks!!!

**Comment by NeilJohnson0930** [comment]
@Wendong-Fan  @lightaime @zhaochenyang20 New  token per second result has been updated, feel free to review

**Pull request merged**